MBX package

Miniland1333 opened the pull request:
**Summary**

Plugin for integrating the [MBX library](https://mbxsimulations.com/) with LAMMPS for running MB-nrg simulations such as with MB-pol. 

**Author(s)**

Plugin authors: @Miniland1333, @cjknight 

MBX authors: @paesanilab, corresponding email Francesco Paesani fpaesani@ucsd.edu

**Licensing**

By submitting this pull request, I agree, that my contribution of this plugin will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

Do note that the MBX library itself retains its separate licensing and is distributed separately. Copyright © 2025 The Regents of the University of California.

**Backward Compatibility**

N/A

**Implementation Notes**

- MBX Plugin is based off of the PLUMED plugin.
- Ran clang to ensure both the documentation and plugin code are formatted correctly.
- Both documentation and code builds complete successfully.

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [x] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

[MBX website](https://mbxsimulations.com/)
[MBX Github](https://github.com/paesanilab/MBX)
[MBX license](https://github.com/paesanilab/MBX/blob/master/LICENSE)

Miniland1333 committed:
Initial LAMMPS+MBX CMake

Miniland1333 committed:
Disabled optimizations

Miniland1333 committed:
Initial MBX plugin documentation

Miniland1333 committed:
Merge branch 'lammps:develop' into MBX-plugin

Miniland1333 committed:
Update MBX version to 1.3.0

Miniland1333 committed:
Updated fix_mbx.cpp to v1.3.0

Miniland1333 committed:
Enabled optimization

Miniland1333 committed:
Performed clang-format

Miniland1333 committed:
Merge branch 'lammps:develop' into MBX-plugin

Miniland1333 committed:
Documentantion1

Miniland1333 committed:
Documentation2

Miniland1333 committed:
Documentation3

Miniland1333 committed:
Merge remote-tracking branch 'upstream/develop' into MBX-plugin

Miniland1333 committed:
Added MBX examples

Miniland1333 committed:
Documentation 5

Miniland1333 committed:
Documentation 6

Miniland1333 committed:
Ran clang-format

Miniland1333 committed:
Add eperm and eind terms printing

Miniland1333 committed:
Changed ordering of mbx_thermo_print

Miniland1333 committed:
Added readme to MBX package

Miniland1333 committed:
Added links to run.sh

Miniland1333 committed:
Merge remote-tracking branch 'upstream/develop' into MBX-plugin

Miniland1333 committed:
Removed a whitespace

Miniland1333 committed:
Applied make check-errordocs fixes

akohlmey commented:
> Plugin for integrating the [MBX library](https://mbxsimulations.com/) with LAMMPS

Please note that in the context of LAMMPS this would be called a "package". For LAMMPS plugins would be compiled separately and loaded dynamically at runtime with the "plugin" command.

akohlmey commented:
> Do note that the MBX library itself retains its separate licensing and is distributed separately. Copyright © 2025 The Regents of the University of California.

Can you please clarify if that license is compatible with the GPLv2 (only) used for LAMMPS?
This doesn't affect the inclusion of the package sources (which *must* be re-licensed as GPLv2 or under a less restrictive, compatible license as indicated), but is of interest to people that package LAMMPS binaries. E.g. a distribution of external sources under a license not compatible with GPLv2 (only), would disallow distributing pre-compiled binaries with the package included. It would have to be converted to a LAMMPS plugin and loaded dynamically at runtime.

akohlmey commented on examples/PACKAGES/mbx/run.sh:
Please remove this script from the examples and the symlinks to it. Instead, examples should be run with `mpirun -np 1` and `mpirun -np 4` and the log file recorded as in the other examples bundled with LAMMPS.

Miniland1333 replied:
Removed script as requested, and added in example log files similarly to other examples.

akohlmey commented on src/MBX/Install.sh:
We are planning to stop support for building packages that require building a bundled or external library with the traditional build system. See pull request #4663. The `Install.sh` in your package should be written in a similar fashion, i.e. supporting only uninstalling the package. This version is incomplete anyway since it is missing the corresponding files in the `lib` folder and the processing of Makefile.packages to include the settings etc.

Miniland1333 replied:
Removed unused `Install.sh` due to the deprecation of the traditional build system.

akohlmey commented on examples/PACKAGES/mbx/256h2o/lammps.in:
Please follow the conventions of other LAMMPS examples to have input file named `in.<name>`, and also remove the block of variables at the beginning and use their values instead, since using the variables makes reading the example file harder.

Miniland1333 replied:
Renamed input files to match other examples, and inlined variables.

akohlmey commented on src/MBX/fix_mbx.h:
Please update `src/.gitignore` with the files from this package.

Miniland1333 replied:
Updated `src/.gitignore` as requested. I wonder if this step is still needed due to the impending deprecation of the traditional build system in favor of CMake.

akohlmey commented on cmake/CMakeLists.txt:
Since you are adding a package, probably some of the files in `cmake/presets` may need to be updated.

Miniland1333 replied:
Updated `all_on.cmake` and `all_off.cmake`.

Since I am currently assuming that MBX should not be part of pre-compiled binaries due to the possible licensing issues, do let me know if there are other presets you would recommend modifying.

Miniland1333 committed:
Rename input files

Miniland1333 committed:
Remove unused make install script

Miniland1333 commented:
Dear @akohlmey, thank you for your detailed review feedback! I will have some responses to your feedback as I go through them.

Here is a summary for myself as I implement fixes to address your comments:
- [x] Confirm license compatibility or incompatibility with pre-compiled binaries
- [ ] [related to license] consider having MBX as a LAMMPS plugin and load dynamically at runtime
- [x] Remove run.sh script and ensure log file is recorded similarly to other examples bundled with LAMMPS
- [X] Remove support for traditional build system
- [X] Rename input file `in.<name>` and inline the variables
- [x] Update `src/.gitignore`
- [x] Modify `cmake/presets`

Feedback by jtclemm:

- [x] Add additional context to `fix_mbx.rst` about the MB-nrg models and the acronyms
- [x] Provide more detailed feedback about the fix arguments
- [x] Add angstrom units for cutoff.
- [x] Add description for pair_coeffs similar to TERSOFF (since it is also many-body)

I will also let you know once I have addressed all of your concerns and am ready for another review.

jtclemm commented on doc/src/fix_mbx.rst:
```diff
@@ -0,0 +1,126 @@
+
+This fix instructs LAMMPS to call the `MBX library <_mbxwebsite>`_
+in order to simulate MB-nrg models such as MB-pol. This fix must be
+used in conjunction with the :doc:`pair mbx <pair_mbx>` command.
```
My personal opinion, I find it very helpful when documentation pages provide at least some context for users not familiar with the exact methodology. What is the MBX library? What do the acronyms MBX, MB-nrg, and MB-pol stand for?

jtclemm commented on doc/src/fix_mbx.rst:
```diff
@@ -0,0 +1,126 @@
+:doc:`lj/cut <pair_lj>`. Do note that all electrostatics must be computed within MBX, so the
+:doc:`coul/exclude <pair_coul>` pair_style must be applied on the non-MB-nrg molecules.
+See  ``examples/PACKAGES/mbx`` for a complete hybrid example.
+
```
Would be helpful to explain keywords/arguments here as in other fix styles.

jtclemm commented on doc/src/pair_mbx.rst:
```diff
@@ -0,0 +1,103 @@
+
+    pair_style mbx cutoff
+
+* cutoff = real-space cutoff for MBX. 9.0 is usually a safe value.
```
9.0 in which units?

jtclemm commented on doc/src/pair_mbx.rst:
```diff
@@ -0,0 +1,103 @@
+:doc:`lj/cut <pair_lj>`. Do note that all electrostatics must be computed within MBX, so the
+:doc:`coul/exclude <pair_coul>` pair_style must be applied on the non-MB-nrg molecules.
+See  ``examples/PACKAGES/mbx`` for a complete hybrid example.
+
```
A description of pair coeffs seems to be missing.

Miniland1333 replied:
Since MBX is many-body, we currently do not use the values for `pair_coeff`. Do you think it would be better for the end-user if we keep the placeholder value `pair_coeff * * 0.0 0.0` in our examples, or would it be clearer to just remove the `pair_coeff` since the values aren't used?

akohlmey replied:
@Miniland1333 with many-body potentials in LAMMPS the pair_coeff command is used to a) specify the name(s) of the parameter file(s) and b) provide a mapping of LAMMPS' (numeric) atom types to the (usually symbolic) parameter element entries. Please see pair styles like tersoff, airebo, meam, or snap to name a few examples.

akohlmey replied:
If you want to ignore any pair_coeff settings, you have to follow the example of pair style zero, where the command would be `pair_coeff * *` without any arguments. The pair_coeff command is required, since it is checked whether there is a complete set of parameters. You can set the `one_coeff` class member to 1, if you only want to allow a single `pair_coeff * *` command.

Miniland1333 replied:
Thanks again for your suggestions about `pair_coeff`!

Short term (this PR)
- Set the default to `pair_coeff * *`
- tested and enabled `one_coeff = 1;`
- tested and declined to add `manybody_flag`.

Long term (not this PR), during some future MBX refactor we'll attempt to make the `pair_coeff` functionality more similar to what `tersoff` is using.

akohlmey replied:
> Long term (not this PR), during some future MBX refactor we'll attempt to make the pair_coeff functionality more similar to what tersoff is using.

I don't think this needs much of a refactor. You could rename fix mbx to fix MBX and make it thus an "internal fix" (undocumented) and create the fix instance as part of the pair_coeff command and all the fix arguments would be the pair_coeff arguments after the  '* *'

akohlmey replied:
> tested and declined to add manybody_flag

Why? Your pair style *should* add it since it is incompatible with molecular systems unless it has the properly inclusive special_bonds settings.

Miniland1333 replied:
For your question about `manybody_flag`, MBX is designed to operate both standalone as well as paired with MD engines such as LAMMPS or I-PI. Therefore, MBX internally always constructs its own neighbor list and does not rely on LAMMP's neighbor list.

Additionally, we do have some hybrid/overlay use cases where say an MBX water molecule is a guest in a classical host system. If hypothetically the host system is using Amber, it would need `special_bonds amber` for that portion of the interaction. However, it sounds like both according to the LAMMPS documentation and [pair.cpp](https://github.com/lammps/lammps/blob/develop/src/pair.cpp#L241-L250) that the `manybody_flag = 1` would override the `special_bonds` settings, which for this use case I worry would cause issues.
If you are curious, here is a concrete example of an [MB-pol water molecule in an AMBER CALF-20 MOF](https://github.com/paesanilab/MBX/blob/d6db1277d7e0aba47408f14ef10a2fe9cd66e176/plugins/lammps/examples/mof_ff_lammps%2B6_h2o_mbx/lammps.in#L146).

Hopefully this makes sense, but of course please let me know if you have clarification questions about this. I personally would like to hold off on adding the `manybody_flag` option until my team has a better understanding of how it would affect these use cases.

------------------------------------


> > Long term (not this PR), during some future MBX refactor we'll attempt to make the pair_coeff functionality more similar to what tersoff is using.
> 
> I don't think this needs much of a refactor. You could rename fix mbx to fix MBX and make it thus an "internal fix" (undocumented) and create the fix instance as part of the pair_coeff command and all the fix arguments would be the pair_coeff arguments after the '* *'

For your comment about the refactor, thanks for the information! I'm starting to test this out on a separate branch, but I'll save this change for a future PR and bundle it with some other related medium-term changes. Your suggestion of using an "internal fix" (undocumented) should be quite useful.

Thanks again for your advice and for taking the time to look at this PR!

akohlmey replied:
> but I'll save this change for a future PR

I consider this a bad idea since that will require a change of the user interface. This is something we try to avoid (unlike internal changes that leave the interface unchanged).

Miniland1333 replied:
FYI I have a working prototype of the internal fix, and I am still working on some related changes updating the documentation. I will let you know once I have completed these tasks and am ready for a review.

- [X] Implemented internal fix
- [x] Update documentation
- [x] Finish adding strong input validation
- [x] Update examples

Miniland1333 commented:
> > Do note that the MBX library itself retains its separate licensing and is distributed separately. Copyright © 2025 The Regents of the University of California.
> 
> Can you please clarify if that license is compatible with the GPLv2 (only) used for LAMMPS? This doesn't affect the inclusion of the package sources (which _must_ be re-licensed as GPLv2 or under a less restrictive, compatible license as indicated), but is of interest to people that package LAMMPS binaries. E.g. a distribution of external sources under a license not compatible with GPLv2 (only), would disallow distributing pre-compiled binaries with the package included. It would have to be converted to a LAMMPS plugin and loaded dynamically at runtime.

Dear @akohlmey,
I am still discussing with my team the implications of the [license](https://github.com/paesanilab/MBX/blob/master/LICENSE).

Hypothetically, in the event that the license is indeed **not** compatible with GPLv2, let me know if I am interpreting your statement correctly:

The MBX package (static linkage) can be in the package sources, but it **_cannot be distributed in pre-compiled binaries_** if it has a GPLv2 incompatibility. Therefore, in that case it would be recommended to also have an MBX plugin (dynamic linkage) that can work with pre-compiled binaries. Users building from source could use MBX package, while users using pre-compiled binaries will need to use the MBX plugin version.

Please let me know if my interpretation is correct or if you are suggesting something different!

Miniland1333 committed:
Merge branch 'develop' into MBX-plugin

Miniland1333 committed:
Merge branch 'MBX-plugin' of https://github.com/Miniland1333/lammps into MBX-plugin

akohlmey commented:
> Please let me know if my interpretation is correct or if you are suggesting something different!

That is how we interpret the situation. There may be some "GPL activists" with a more radical interpretation, but that is not how *we* roll.

Miniland1333 committed:
Added MBX cmake preset

Miniland1333 committed:
Add MBX to gitignore

Is this still useful in a cmake-only build process?

Miniland1333 committed:
Added MBX log files

Miniland1333 committed:
Updated example

Miniland1333 committed:
Documentation 7

Miniland1333 committed:
Update MBX pair_coeff handling and input files

Modified MBX input examples to remove unused pair_coeff parameters. Updated pair_mbx.cpp to accept fewer arguments for pair_coeff.

akohlmey committed:
We should still have an Install.sh file in MBX so we get a suitable error.

If there is no explicit Install.sh in the package. The default Install.sh
file is used and that would still (incorrectly) allow doing `make yes-mbx`.

Miniland1333 committed:
Cleaned up pair_coeff flags

Miniland1333 committed:
Merge remote-tracking branch 'upstream/develop' into MBX-plugin

Miniland1333 committed:
Documentation 8

Miniland1333 committed:
Updated MBX log files

Miniland1333 commented:
Hopefully I have addressed all of your previous feedback @akohlmey and @jtclemm, so this PR should be ready for a re-review.

Based on our previous conversations, one thing to note is that we are assuming MBX has license incompatibility with GPLv2 and thus ***cannot be distributed in pre-compiled binaries***. Please let me know if I need to make any additional changes to comply with this restriction.

akohlmey committed:
documentation fixup: references, non-ASCII characters, spelling

Miniland1333 committed:
Merge branch 'develop' into MBX-plugin

Miniland1333 committed:
Make fix_mbx internals as fix_MBX

Miniland1333 committed:
Merged fix_mbx docs into pair_mbx

Miniland1333 committed:
Merge branch 'develop' into MBX-plugin

Miniland1333 committed:
Enhance MBX fix input validation and error reporting

- Added MBXParseResult struct for better error handling.
- Implemented reportFailure and test_MBX_fix_parsing methods to validate input arguments.
- Updated error messages in FixMBX to include "[MBX]" prefix for consistency.
- Removed deprecated print/dipoles option from documentation.
- TODO: Add support back in for dp1 monomer

Miniland1333 committed:
Strengthened MBX input validation

Miniland1333 committed:
Added MBX dipole example

Miniland1333 committed:
Modified MBX_dipoles example frequency

Miniland1333 committed:
MBX clang-format

Miniland1333 committed:
Removed newline

Miniland1333 committed:
Added special_bonds example

Miniland1333 committed:
Added more water examples

Miniland1333 committed:
Added Ewald warnings

Miniland1333 committed:
Documentation 9

Miniland1333 commented:
@akohlmey and @jtclemm we have made significant improvements to the package and it is now ready for another review!

Here is a brief summary of the recent modifications:
- Made fix_MBX an internal fix that is called by pair_mbx
- Implemented the new interface involving passing arguments through `pair_coeff`
- Updated documentation and added additional examples
- Added significantly more input validation for the pair_coeffs line

I think your suggestion of having the simplified internal fix is indeed more intuitive and user-friendly, so I'm glad we've now implemented it.

Thanks in advance for looking over this PR again, and of course please let me know if you have any questions or concerns.

Miniland1333 committed:
Added support for LAMMPS minimization

Miniland1333 committed:
Added version printing

Miniland1333 committed:
Ran clang-format again

Miniland1333 committed:
Merge branch 'develop' into MBX-plugin

Miniland1333 committed:
Updated dipole example

Miniland1333 commented:
@akohlmey just a brief reminder that this PR is ready for review whenever you have time!

jtclemm commented on src/MBX/pair_mbx.cpp:
```diff
@@ -0,0 +1,772 @@
+
+  fix_args = fmt::format("_FIX_MBX_INTERNAL all MBX {}", fix_args);
+
+  fix_MBX = dynamic_cast<FixMBX *>(modify->add_fix(fix_args));
```
Why create the fix in the coeff method? What happens if the user calls the pair_coeff command twice?

akohlmey replied:
@jtclemm I suggested this since this made the pair style more similar to other pair styles and avoids having to create a fix explicitly.

When the pair_coeff statement is given a second time, it is like issuing a fix command with the same fix-ID a second time, which will update the fix with the new settings. This is in line with expected behavior. It is also possible to program this more defensively and check for existence of the fix and delete and re-create it.

I have not had the time to properly review the actual implementation details after the significant revisions that were done.

jtclemm replied:
I see. I would think passing the args for the fix as settings and then initializing the fix there would make more sense. Some example files have lots of `pair_coeff` calls which seems inefficient (and possibly confusing if the fix args differ between some `pair_coeff` calls).

akohlmey replied:
> Some example files have lots of pair_coeff calls which seems inefficient

This pair style is a many-body potential and for those it is required to have only one pair_coeff statement with \* \* as arguments and then the potential file and some form of atom type to element mappings. Please note the `one_coeff = 1;` line in the constructor.

jtclemm replied:
Ah I missed that. I was looking at in.mbx_mof_1h2o, but I see now that those calls are all for lj/cut or buck.

jtclemm commented on src/MBX/pair_mbx.cpp:
```diff
@@ -0,0 +1,772 @@
+  // if (strcmp(str,"epsilon") == 0) return (void *) epsilon;
+  // if (strcmp(str,"sigma") == 0) return (void *) sigma;
+  return NULL;
+}
```
Why not delete all these unused methods to simplify the code? Ditto for some methods in the fix (e.g. reverse comm).

Miniland1333 replied:
Thanks for pointing these out! I've removed both some unused functions and some unused commented-out code in those files. I'm taking another look through, and if there are any additional unused methods you see please let me know!

jtclemm commented on doc/src/pair_mbx.rst:
```diff
@@ -0,0 +1,203 @@
+        compute         mbx all pair mbx
+
+The *json* argument specifies the name of the MBX JSON configuration
+file to use, such as `mbx.json`.  If this file is not provided, the fix
```
Should this be, "the pair style will attempt..."?

Miniland1333 replied:
You are correct! I have made this change!

Miniland1333 committed:
Merge branch 'develop' into MBX-plugin

Miniland1333 committed:
Updated documentation

akohlmey commented:
@Miniland1333 I should have time to take a closer look at your pull request in the next days. I already saw a few more changes that I would like to have made. So far I have not even compiled the package, so there are likely issues. We can do this in two different ways.

1. I can just make the changes myself and then get them integrated into the pull request, or
2. I can give comments and suggest changes through GitHub.

The choice 1. is easier for me to do, but you will have to trust me that I am not messing up things too much or implement changes that you do not want. The choice 2. is more educational but also more work for both of us. Please let me know.

Miniland1333 commented:
> @Miniland1333 I should have time to take a closer look at your pull request in the next days. I already saw a few more changes that I would like to have made. So far I have not even compiled the package, so there are likely issues. We can do this in two different ways.
> 
> 1. I can just make the changes myself and then get them integrated into the pull request, or
> 2. I can give comments and suggest changes through GitHub.
> 
> The choice 1. is easier for me to do, but you will have to trust me that I am not messing up things too much or implement changes that you do not want. The choice 2. is more educational but also more work for both of us. Please let me know.

Since it sounds like there are multiple changes that you would like to request, let's go with choice 2 (Github change suggestions) since that should be cleaner for me to learn from and validate.

For your compile, I have two notes that may be useful:
1. For development purposes only, in MBX.cmake I would recommend adding `--disable-optimization` since otherwise the optimized compile can take a long time.
    1. ```
       CONFIGURE_COMMAND autoreconf -fi && <SOURCE_DIR>/configure
                                            --prefix=<INSTALL_DIR>
                                            --disable-optimization
                                            ${MBX_CONFIG_MPI}
                                            CXX=${MBX_CONFIG_CXX}
                                            CC=${MBX_CONFIG_CC}
       ```
2. During the compilation, I do see a `NLOHMANN_JSON_VERSION` warning. We're updating our NLOHMANN version to match yours, but perhaps we should also add some `#ifdef/#ifndef` to protect against these version clashes.

Thank you both for your time and for being so thorough!

Miniland1333 committed:
Remove unused pair_mbx functions

Miniland1333 committed:
Remove unused fix_mbx functions

akohlmey commented on cmake/Modules/Packages/MBX.cmake:
```diff
@@ -0,0 +1,118 @@
+endif()
+
+
+set(MBX_URL "https://github.com/paesanilab/MBX/archive/refs/tags/v1.3.0.tar.gz"
```
Rather than referring to a git tag and let GitHub do the archive creation, it would be better to upload a proper tar archive yourself and have a proper release, too.

Some time ago, the GitHub admins changed the compression program and that broke all checksums for such auto-generated download. They reverted the change, but with an explicit upload you can avoid this issue from the beginning.

Please also see my comment on `CONFIGURE_COMMAND`.

akohlmey replied:
P.S.: For LAMMPS and LAMMPS-GUI we have taken this a step farther and use a recently added GitHub feature to make any formal releases "immutable", i.e. all uploads must be completed *before* making the release and afterwards neither the tag nor the uploads may be changed and are cryptographically confirmed. This is to reduce the attack vector for supply-chain attacks. Granted, that scientific software is not a primary target, since its audience is small, but since the LAMMPS audience seems to be growing continuously for the last 15+ years, we may become an increasingly attractive target.

akohlmey commented on cmake/Modules/Packages/MBX.cmake:
```diff
@@ -0,0 +1,118 @@
+set(MBX_SUFFIX "" CACHE STRING "Suffix for MBX library")
+mark_as_advanced(MBX_SUFFIX)
+
+if(BUILD_MPI)
```
This whole block could be simplified:
```
  diff --git a/cmake/Modules/Packages/MBX.cmake b/cmake/Modules/Packages/MBX.cmake
  index 2c33cbeee6..5e6982ad6c 100644
  --- a/cmake/Modules/Packages/MBX.cmake
  +++ b/cmake/Modules/Packages/MBX.cmake
  @@ -10,26 +10,14 @@ endif()
   set(MBX_SUFFIX "" CACHE STRING "Suffix for MBX library")
   mark_as_advanced(MBX_SUFFIX)
   
  +set(MBX_CONFIG_CC  ${CMAKE_C_COMPILER})
  +set(MBX_CONFIG_CXX  ${CMAKE_CXX_COMPILER})
   if(BUILD_MPI)
     set(MBX_CONFIG_MPI "--enable-mpi")
  -  set(MBX_CONFIG_CC  ${CMAKE_C_COMPILER})
  -  set(MBX_CONFIG_CXX  ${CMAKE_CXX_COMPILER})
  -  # set(MBX_CONFIG_CPPFLAGS "-I ${MPI_CXX_INCLUDE_PATH}")
  -  # set(MBX_CONFIG_LIB "${MPI_CXX_LIBRARIES}")
  -  # set(MBX_CONFIG_DEP "")
   else()
     set(MBX_CONFIG_MPI "--disable-mpi")
  -  set(MBX_CONFIG_CC  ${CMAKE_C_COMPILER})
  -  set(MBX_CONFIG_CXX  ${CMAKE_CXX_COMPILER})
  -  # set(MBX_CONFIG_CPPFLAGS "")
  -  # set(MBX_CONFIG_LIB "")
  -  # set(MBX_CONFIG_DEP "")
  -
  -
  -
   endif()
   
  -
```

akohlmey commented on cmake/Modules/Packages/MBX.cmake:
```diff
@@ -0,0 +1,118 @@
+
+mark_as_advanced(MBX_URL)
+mark_as_advanced(MBX_MD5)
+GetFallbackURL(MBX_URL MBX_FALLBACK)
```
Where is MBX_FALLBACK defined?
This would usually be pointing to the LAMMPS download server (so we can have a clean copy, even if for some reason GitHub won't be accessible or changes the downloads again).
```suggestion
```

akohlmey commented on cmake/Modules/Packages/MBX.cmake:
```diff
@@ -0,0 +1,118 @@
+  ExternalProject_Add(mbx_build
+    URL     ${MBX_URL} ${MBX_FALLBACK}
+    URL_MD5 ${MBX_MD5}
+    BUILD_IN_SOURCE 1
```
Requiring to build in the source tree is almost always a bad idea and thus something that should be changed in the external project.

akohlmey commented on cmake/Modules/Packages/MBX.cmake:
```diff
@@ -0,0 +1,118 @@
+    URL     ${MBX_URL} ${MBX_FALLBACK}
+    URL_MD5 ${MBX_MD5}
+    BUILD_IN_SOURCE 1
+    CONFIGURE_COMMAND autoreconf -fi && <SOURCE_DIR>/configure
```
Requiring the user compiling LAMMPS to have the full autoconf toolchain installed so that autoreconf works is putting the burden pretty high. I would consider it a significant improvement to provide those created files with the source package (and that is the general idea for autotools). Only developers using the source code management to follow the development should be required to have those tools installed.

As far as LAMMPS is concerned this drops the chances to have the package pre-built as a plugin close to zero.

Miniland1333 committed:
Merge branch 'develop' into MBX-plugin

akohlmey commented on doc/src/pair_mbx.rst:
```diff
@@ -0,0 +1,203 @@
+.. code-block:: LAMMPS
+
+    # For a system involving water (atom types O=1, H=2)
+    processors      * * * map xyz
```
The processors keyword should not be needed to demonstrate the usage of the pair style.
Same applies to the compute command. Also the comments should be removed here and we don't as many examples here. This part of the documentation should just show the syntax, so one plain example and one with hybrid should suffice. The explanations in the comments belong to the example input decks to which there is a reference given below.

akohlmey commented on doc/src/pair_mbx.rst:
```diff
@@ -0,0 +1,203 @@
+    models that may also use geometrically dependent charges, such as
+    the Partridge and Schwenke charges used in MB-pol water.
+
+    Therefore, one should never use a coulombic pair style in LAMMPS
```
```suggestion
    Therefore, one should **never** use a coulombic pair style in LAMMPS
```

akohlmey commented on doc/src/pair_mbx.rst:
```diff
@@ -0,0 +1,203 @@
+    the Partridge and Schwenke charges used in MB-pol water.
+
+    Therefore, one should never use a coulombic pair style in LAMMPS
+    such as `coul/cut` or `coul/long` when also using MBX. This mistake
```
```suggestion
    such as `coul/cut` or `coul/long` when also using MBX. This
```

akohlmey commented on examples/PACKAGES/mbx/1024h2o/in.mbx_h2o:
```diff
@@ -0,0 +1,58 @@
+pair_style      mbx 9.0
+pair_modify     mix arithmetic
+
+bond_style      none
```
Undefined force styles default to none. No need to make this explicit.
```suggestion
```
Same for other cases.

akohlmey commented on examples/PACKAGES/mbx/1024h2o/in.mbx_h2o:
```diff
@@ -0,0 +1,58 @@
+
+restart         10 restart.1 restart.2
+
+run             100  upto
```
The upto keyword is redundant here and usually not used in LAMMPS unless required.
```suggestion
run             100
```

akohlmey commented on examples/PACKAGES/mbx/1024h2o/in.mbx_h2o:
```diff
@@ -0,0 +1,58 @@
+fix             thermo_print all print 1 "$(step) $(0.001*time) $(temp) $(etotal) $(ke) $(pe) $(enthalpy) $(density) $(lx) $(ly) $(lz) $(vol) $(pxx) $(pyy) $(pzz) $(press)" file thermo.out screen no title "# step time temp etotal ke pe enthalpy density lx ly lz vol pxx pyy pzz press"
+fix             verbose_mbx_thermo_print all print 1 "$(step) $(0.001*time) $(temp) $(etotal) $(ke) $(pe) $(enthalpy) $(density) $(lx) $(ly) $(lz) $(vol) $(pxx) $(pyy) $(pzz) $(press) $(v_e1bpip) $(v_e2bpip) $(v_e3bpip) $(v_e4bpip) $(v_edisp) $(v_ebuck) $(v_eperm) $(v_eind) $(v_eele) $(v_etot)" file mbx_thermo.out screen no title "# step time temp etotal ke pe enthalpy density lx ly lz vol pxx py ...
+
+dump            1 all custom 1 dump.lammpstrj id mol type q x y z
```
All commands producing output files (dump, restart, write_data, write_restart) should be commented out to avoid flooding the examples tree with unwanted files when we use it for testing.

akohlmey commented on examples/PACKAGES/mbx/1024h2o/in.mbx_h2o:
```diff
@@ -0,0 +1,58 @@
+processors      * * * map xyz
```
Not really needed
```suggestion
```

akohlmey commented on examples/PACKAGES/mbx/1024h2o/in.mbx_h2o:
```diff
@@ -0,0 +1,58 @@
+boundary        p p p
+
+read_data       initial.data 
+#read_restart    restart.old
```
```suggestion
```

akohlmey commented on examples/PACKAGES/mbx/1024h2o/in.mbx_h2o:
```diff
@@ -0,0 +1,58 @@
+#read_restart    restart.old
+
+pair_style      mbx 9.0
+pair_modify     mix arithmetic
```
There does not seem to be a pair style present that supports mixing. Why use the command when it has no effect? It just confuses beginners to think that it would be needed.

```suggestion
```

akohlmey commented on examples/PACKAGES/mbx/1024h2o/in.mbx_h2o:
```diff
@@ -0,0 +1,58 @@
+thermo          1
+thermo_modify   flush yes
+
+fix             thermo_print all print 1 "$(step) $(0.001*time) $(temp) $(etotal) $(ke) $(pe) $(enthalpy) $(density) $(lx) $(ly) $(lz) $(vol) $(pxx) $(pyy) $(pzz) $(press)" file thermo.out screen no title "# step time temp etotal ke pe enthalpy density lx ly lz vol pxx pyy pzz press"
```
Why repeat the thermo output? Just wastes disk space.

akohlmey commented on examples/PACKAGES/mbx/1024h2o/in.mbx_h2o:
```diff
@@ -0,0 +1,58 @@
+thermo_modify   flush yes
+
+fix             thermo_print all print 1 "$(step) $(0.001*time) $(temp) $(etotal) $(ke) $(pe) $(enthalpy) $(density) $(lx) $(ly) $(lz) $(vol) $(pxx) $(pyy) $(pzz) $(press)" file thermo.out screen no title "# step time temp etotal ke pe enthalpy density lx ly lz vol pxx pyy pzz press"
+fix             verbose_mbx_thermo_print all print 1 "$(step) $(0.001*time) $(temp) $(etotal) $(ke) $(pe) $(enthalpy) $(density) $(lx) $(ly) $(lz) $(vol) $(pxx) $(pyy) $(pzz) $(press) $(v_e1bpip) $(v_e2bpip) $(v_e3bpip) $(v_e4bpip) $(v_edisp) $(v_ebuck) $(v_eperm) $(v_eind) $(v_eele) $(v_etot)" file mbx_thermo.out screen no title "# step time temp etotal ke pe enthalpy density lx ly lz vol pxx py ...
```
The variables can be added to the thermo output and thus make this second fix print superfluous. Also it has more redundant data and thus wastes even more diskspace.

akohlmey commented on examples/PACKAGES/mbx/2048h2o/in.mbx_h2o:
```diff
@@ -0,0 +1,58 @@
+processors      * * * map xyz
```
See my comments on the previous example.

akohlmey replied:
Why do we need four different size water examples with identical inputs?
Just use the smallest and be done with it.

akohlmey commented on examples/PACKAGES/mbx/1024h2o/log.24Sep25.1024h2o.g++.1:
```diff
@@ -0,0 +1,317 @@
+  3072 atoms
+  reading velocities ...
+  3072 velocities
+  scanning bonds ...
```
Why a data file with bonds and angles, when there is only a pair style?
This has an impact on the neighbor list unless `special_bonds lj/coul 1.0 1.0 1.0` is used. Is this desired? Please see also the warning below.

akohlmey commented on examples/PACKAGES/mbx/1024h2o/log.24Sep25.1024h2o.g++.1:
```diff
@@ -0,0 +1,317 @@
+      pair build: half/bin/newton
+      stencil: half/bin/3d
+      bin: standard
+WARNING: Inconsistent image flags
```
This needs be investigated. With a properly prepared data file for just water molecules, this warning should not happen.

akohlmey commented on src/MBX/README:
```diff
@@ -0,0 +1,22 @@
+This package implements the "fix mbx" and "pair_style mbx" commands
```
The fix is now internal.
```suggestion
This package implements the "pair_style mbx" command
```

akohlmey commented on src/MBX/fix_mbx.h:
```diff
@@ -0,0 +1,220 @@
+/* -*- c++ -*- ----------------------------------------------------------
```
Even if the fix style name is uppercase, the file names should be all lowercase. See fix _respa.h for example.

akohlmey commented on src/MBX/fix_mbx.h:
```diff
@@ -0,0 +1,220 @@
+
+// MBX
+
+#include "bblock/system.h"
```
There should be no MBX library includes in the header files.
If specific data structures or types are needed, then the PIMPL programming pattern should be used where there is a pointer to a struct for them. For that struct pointer
there can be a forward declaration and the full implementation is in the .cpp file.

see the use of `struct ACEImpl` in pair style pace as an example.

akohlmey commented on src/MBX/fix_mbx.h:
```diff
@@ -0,0 +1,220 @@
+
+#include "bblock/system.h"
+
+enum {
```
It is a bad idea to define this enum outside the LAMMPS_NS namespace and also outside the fix class definition. It invites namespace clashes. Can this be moved to the .cpp file? If it needs to be shared with the pair style, then move it to a separate header and include it by either style's .cpp file.

Miniland1333 replied:
@akohlmey I have now nested the enum inside of `FixMBX` in fix_mbx.h. https://github.com/lammps/lammps/pull/4694/commits/ebde8b86c0eb1af008f535d17cb73e2f2ba54a91

Please let me know if this is acceptable or if you would insist on moving it to a separate header that is then included by fix_mbx.cpp and pair_mbx.cpp.

akohlmey commented on src/MBX/fix_mbx.h:
```diff
@@ -0,0 +1,220 @@
+  MBXT_NUM_TIMERS
+};
+
+struct MBXParseResult {
```
same issue as with the enum. This is polluting the namespace and it doesn't look like it is needed in the header. We want the header files to be as bare as possible since the Force class has to include all of them (for pairs, bonds, angles, dihedrals, impropers, and kspace, that is) at the same time and anything that is not protected by namespaces can lead to clashes, but not having it in the header is even better.

akohlmey commented on src/MBX/fix_MBX.h:
```diff
@@ -0,0 +1,220 @@
+
+  double **mbx_dip;
+
+  // rank 0's copy of all atoms in simulation cell
```
Is this comment true?? Does the fix have to hold a copy of the entire system? This would be a *very big* problem as this seriously limits the use of this pair style inside LAMMPS and invalidates all the effort in LAMMPS to distribute the per-atom data across subdomains. I would have to talk to the other LAMMPS developers, if we want to include a package with such a limitation.

Miniland1333 replied:
I have now refactored out these older "full" variables and methods that were not MPI-parallelized. We now just use the newer "local" implementation that distributes the data across subdomains.

akohlmey commented on src/MBX/pair_mbx.h:
```diff
@@ -0,0 +1,80 @@
+#ifndef LMP_PAIR_MBX_H
+#define LMP_PAIR_MBX_H
+
+#include "fix_MBX.h"
```
This include should not be used here, but instead a forward declaration of the fix class. See my related comments in the fix class header file.

akohlmey commented on src/MBX/pair_mbx.h:
```diff
@@ -0,0 +1,80 @@
+
+// MBX
+
+#include "bblock/system.h"
```
Same thing about library specific headers as in the fix class header.

akohlmey commented on src/MBX/pair_mbx.h:
```diff
@@ -0,0 +1,80 @@
+#include "bblock/system.h"
+
+
+//TODO Re-enable this once MBX is properly versioned at >1.3.2
```
See above about including library header files.

akohlmey commented on src/MBX/pair_mbx.cpp:
```diff
@@ -0,0 +1,751 @@
+------------------------------------------------------------------------- */
+
+#include "pair_mbx.h"
+#include "atom.h"
```
```suggestion

#include "atom.h"
```

akohlmey commented on src/MBX/pair_mbx.cpp:
```diff
@@ -0,0 +1,751 @@
+#include "neighbor.h"
+#include "respa.h"
+#include "update.h"
+#include <cmath>
```
```suggestion

#include <cmath>
```

akohlmey commented on src/MBX/pair_mbx.cpp:
```diff
@@ -0,0 +1,751 @@
+#include "update.h"
+#include <cmath>
+#include <cstring>
+#include <mpi.h>
```
```suggestion
```

akohlmey commented on src/MBX/pair_mbx.cpp:
```diff
@@ -0,0 +1,751 @@
+#include <cstring>
+#include <mpi.h>
+
+#include "domain.h"
```
This line should be moved up to before the `#include "error.h"` statement.

akohlmey commented on src/MBX/pair_mbx.cpp:
```diff
@@ -0,0 +1,751 @@
+
+PairMBX::PairMBX(LAMMPS *lmp) : Pair(lmp)
+{
+  respa_enable = 1;
```
This is incorrect. There are no `compute_inner()`, `compute_middle()`, or `compute_output()` methods implemented which setting this flag to 1 indicates.
```suggestion
```

akohlmey commented on src/MBX/pair_mbx.cpp:
```diff
@@ -0,0 +1,751 @@
+  restartinfo = 0;
+  no_virial_fdotr_compute = 1;
+  one_coeff = 1;
+
```
This setting is missing since the default is 1 and there is no `single()` method:
```suggestion
single_enable = 0;

```

akohlmey commented on src/MBX/pair_mbx.cpp:
```diff
@@ -0,0 +1,751 @@
+  if (narg < 2) error->all(FLERR, "Incorrect num args for pair coefficients");
+  if (!allocated) allocate();
+
+  int ilo, ihi, jlo, jhi;
```
This is redundant. The check for `one_coeff = 1;` is already done in the `Input` class before calling `Pair::coeff()`, thus you could just explicitly loop from 1 to `atom->ntypes`. Also, since there is only the global cutoff, there should not be a need to store and maintain a per-type-pair cutoff.

akohlmey commented on src/MBX/pair_mbx.cpp:
```diff
@@ -0,0 +1,751 @@
+void PairMBX::settings(int narg, char **arg)
+{
+
+  //TODO Re-enable this once MBX is properly versioned at >1.3.2
```
We prefer to not have output whenever a pair style is defined. That can be messy with restarts. If you absolutely have to have an output option, I suggest to have a 'verbose yes/no' keyword added to the `pair_style` command and then output a version line in the `Pair::init_style()` function, if verbose is requested (default is best set to *no*, but that is ultimately your choice).

akohlmey commented on src/MBX/pair_mbx.cpp:
```diff
@@ -0,0 +1,751 @@
+    if (((Respa *) update->integrate)->level_middle >= 0) respa = 2;
+  }
+
+  // currently request neighbor list, but we don't use it for anything...
```
If you don't use a neighbor list, don't request it. You'll save time.
Also the API you are using is obsolete.

The rest of the content of this function can also be removed as it serves no purpose.

akohlmey commented on src/MBX/pair_mbx.cpp:
```diff
@@ -0,0 +1,751 @@
+
+double PairMBX::init_one(int i, int j)
+{
+  if (setflag[i][j] == 0) cut[i][j] = mix_distance(cut[i][i], cut[j][j]);
```
Since all cutoffs are the same, this serves no purpose.

akohlmey commented on src/MBX/pair_mbx.cpp:
```diff
@@ -0,0 +1,751 @@
+          f[ii][1] -= grads[indx++];
+          f[ii][2] -= grads[indx++];
+        }
+#ifndef _DEBUG_EFIELD
```
Is this define really needed?
It looks strange to *add* code when *not* debugging.

akohlmey reviewed (changes requested):
Here is a first batch of suggestions from just reading through the pull request without even trying to compile.

It looks as if the fix holds the data for the entire system. This is a very serious problem. It limits what size system can be handled and also restricts parallel scaling. LAMMPS is designed to avoid this.

Miniland1333 committed:
Removed older "full non-MPI implementation

Miniland1333 committed:
Merge branch 'lammps:develop' into full_refactor

Miniland1333 committed:
mbx summary moved behind print/verbose option.

mbx summary is empty if using non-MPI.

Miniland1333 committed:
Merge branch 'lammps:develop' into full_refactor

Miniland1333 commented:
@akohlmey thanks for your feedback! For your primary concern about how the fix data is handled, I've spent the weekend and refactored out the legacy "full" implementation that did not take advantage of MPI subdomains properly. The newer "local" implementation does distribute the system across the ranks and is now used for both MPI and non-MPI compiles. I'm still doing some testing to make sure I'm not introducing any new bugs.

As for all your other suggestions, I believe I agree with pretty much all of them and will start implementing/testing them.

Miniland1333 committed:
Apply documentation suggestions from code review

1. bold "never"
2. remove "mistake"
3. Readme documentation

Co-authored-by: Axel Kohlmeyer <akohlmey@gmail.com>

Miniland1333 commented:
For my personal reference, here is my outstanding task list based on the previous review:
- [x] Manually add TAR to releases
- [ ] Consider making releases "immutable"
- [x] Simplify BUILD_MPI in MBX.cmake
- [x] Remove fallback url
- [x] Remove BUILD_IN_SOURCE
- [x] Commit autotools-generated files (`./configure` etc.) to MBX as part of new releases
- [ ] Check if "map xyz" can be removed from the processors command
- [X] Corrections to documentation
- [X] Remove "none" bond/angle/dihedral styles
- [X] Remove upto
- [X] Comment out output files (dump, restart, write_data, write_restart) in MBX examples
- [X] Remove restart.old
- [X] Remove pair_modify mix arithmetic
- [X] Remove extra thermo output
- [X] Remove larger water examples
- [X] Data file bonds and angles
- [X] Warning inconsistent image flags
- [X] Readme documentation
- [x] fix_MBX.cpp to fix_mbx.cpp
- [x] remediate bblock/system.h in fix_MBX.h
- [ ] Move enums from fix_MBX.h to fix_MBX.cpp
- [x] struct MBXParseResult polluting namespace
- [x] Remove legacy "full" implementation
- [x] Forward declaration of fix_MBX.h in pair_mbx.h
- [x] remediate bblock/system.h in pair_mbx.h
- [x] remediate mbx_config.h in pair_mbx.h
- [X] Newline before "atom.h" and <cmath> in pair_mbx.cpp
- [X] Remove <mpi.h> from pair_mbx.cpp
- [X] "domain.h" include in pair_mbx.cpp
- [X] Remove respa_enable flag
- [X] single_enable = 0
- [X] Remove redundant for loop in pair_mbx.cpp
- [x] Move version print statement
- [X] Remove unused neighbor list
- [X] Remove cutoff mixing
- [x] See if ifndef _DEBUG_EFIELD can be removed

Miniland1333 committed:
Add SHAKE to example

Miniland1333 committed:
Update MBX examples

Miniland1333 committed:
pair_mbx global cutoff

Miniland1333 committed:
Update MBX example logfiles

Miniland1333 committed:
Remove larger water examples

Miniland1333 committed:
Renamed fix_MBX.* to fix_mbx.*

Miniland1333 committed:
Move version printing to citeme

Miniland1333 committed:
Initial MBXImpl implementation

Miniland1333 committed:
Make json_file a stack string

Miniland1333 committed:
Forward declaration for FixMBX

Miniland1333 committed:
Merge branch 'lammps:develop' into MBX-plugin

Miniland1333 committed:
Simplify MBX.cmake

Miniland1333 committed:
Update setflag loops

Miniland1333 committed:
Nest enum in FixMBX

Miniland1333 committed:
Remove _DEBUG_EFIELD

Miniland1333 committed:
Remove duplicate "the" in Build_extras

Miniland1333 committed:
Merge branch 'develop' into MBX-plugin

Miniland1333 committed:
Merge branch 'develop' into MBX-plugin

Miniland1333 committed:
Added optional MBX_GIT_TAG build option

Miniland1333 committed:
Revert "Added optional MBX_GIT_TAG build option"

This reverts commit 7946e2a6caf58bdc8ee4a5273bbb0d16bebf8a62.

Miniland1333 committed:
Merge remote-tracking branch 'upstream/develop' into MBX-plugin

Miniland1333 committed:
Updated MBX tarball

TODO: Update tarball and checksup before merging!!!

Miniland1333 committed:
Merge branch 'develop' into MBX-plugin

Miniland1333 commented:
Hello @akohlmey! I have made changes that aims to address your feedback. I have accepted all of your recommendations except for "map xyz".

For "map xyz", our current internal implementation of kspace relies on the XYZ ordering of MPI ranks, and we are not confident that fully compatible with the default `map cart` behavior of LAMMPS. We have long-terms plans to rewrite this implementation to be more flexible, but it will be a while before we can get to that.

The plugin currently compiles and runs correctly for me, so I assume the next step will be to have you compile it as well. I see you already resolved an issue I was seeing with the Nlohmann JSON name clashes, so thanks for doing that!

Thanks again for your help with this PR so far, and let me know if anything comes up.

akohlmey commented:
> For "map xyz", our current internal implementation of kspace relies on the XYZ ordering of MPI ranks, and we are not confident that fully compatible with the default `map cart` behavior of LAMMPS. We have long-terms plans to rewrite this implementation to be more flexible, but it will be a while before we can get to that.

Then there should be a warning block at a prominent location in the documentation and ideally some kind of check added that confirms that the correct mapping is used, so that users won't be surprised if they do not set the processor map as required.

I won't be able to have a closer look for a while (probably until after Christmas break), since some other tasks have higher priority at the moment.

Miniland1333 committed:
Updated MBX checksum

Miniland1333 committed:
Updated MBX examples

Miniland1333 committed:
Added processors XYZ to documentation

Miniland1333 committed:
Merge branch 'develop' into MBX-plugin

Miniland1333 committed:
Added example processors error

Miniland1333 committed:
Fix typos found by copilot

Miniland1333 committed:
Merge branch 'develop' into MBX-plugin

Miniland1333 committed:
Added some inline documentation

Miniland1333 committed:
Removed unused code

Miniland1333 committed:
Merge branch 'MBX-plugin' of https://github.com/Miniland1333/lammps into MBX-plugin

Miniland1333 committed:
Added comments to map xyz

Miniland1333 commented:
Okay I have updated the documentation and confirmed that our check for `map xyz` is still functioning as intended!https://github.com/lammps/lammps/blob/b118d2deea9db3178998e73165e4ac4dca5e7ab3/src/MBX/fix_mbx.cpp#L452-L475

Since you appear to be busy at the moment, could you please have Github Copilot review this PR as well, so that I can address its comments while I wait for your review?

This plugin will be very useful for a grant proposal we have submitted, so let me know if there is anything else on my end that I can do to help move things along. @fpaesani this is the PR that is related to the grant proposal.

Miniland1333 committed:
Added a link to the docs

akohlmey commented:
> Since you appear to be busy at the moment, could you please have Github Copilot review this PR as well, so that I can address its comments while I wait for your review?

Done. Be careful with its comments. I've recently seen some reviews from Copilot with bad suggestions. It seems to work most reliable when reviewing documentation, but is struggling to "understand" some of the peculiarities of LAMMPS' code design (which is not conventional due to its roots in Fortran which was converted to "C with classes" and only a few years back moved to a more C++ style of programming, see https://doi.org/10.5281/zenodo.17117558 for a discussion)

Miniland1333 committed:
Delete cmake_recompile.sh

Miniland1333 committed:
Fix whitespace error

Miniland1333 committed:
Updated MBX build tarball to production

Miniland1333 committed:
Merge branch 'develop' into MBX-plugin

Miniland1333 commented:
Dear LAMMPS developers,
    I hope you had a productive winter break. Since the Copilot review came back without any comments, we can continue the review process and hopefully wrap this PR up soon. I have updated the MBX tarball to our latest release and it seems to compile fine!

What do we still need to do in order to wrap up?

fpaesani commented:
Hi @akohlmey,

Following up on our note from last week: we have updated the bundled MBX tarball to the latest upstream release and confirmed it still builds cleanly. Could you let us know if there is anything else you would like us to address before this PR can be approved/merged?

This integration is an important deliverable for our NSF project. Also, two recent benchmarks on arXiv that assess water near melting find MB-pol/MB-pol–trained models provide the best agreement with experiment among the first-principles/MLP approaches they test, suggesting strong community interest in having a robust MBX+LAMMPS integration:
- https://arxiv.org/abs/2512.23939
- https://arxiv.org/abs/2512.23940

Thanks again,
@Miniland1333, @cjknight, and @fpaesani

akohlmey commented:
Hi @fpaesani 

I cannot comment on the review status since I was busy with other projects and will still be too busy for a while.  Thus, a timely continuation of the review and test process will require that some other LAMMPS core developer takes over (or multiple of them).

This does not only apply to your contributions but several that have accumulated over the last few months. Can you let us know if there is any hard deadline until which you would need this feature to be integrated and released? I don't see a principal problems it is just a matter going through the all the steps and making sure that everything is consistent with the rest and compliant with what the LAMMPS developers have agreed should be applied to new contributions.
If we know about important deadlines, we can try to shift priorities, if needed.

fpaesani commented:
Hi @akohlmey,

Thank you for following up. If we could get the MBX integration wrap up by Jan 31, that would be awesome!

Thank you so much for your help and please let us know if there is anything we can do to facilitate this.
@Miniland1333 @cjknight @fpaesani

Miniland1333 committed:
Merge remote-tracking branch 'upstream/develop' into MBX-plugin

Miniland1333 committed:
Change versionadded to TBD

Miniland1333 committed:
Merge branch 'develop' into MBX-plugin

ndtrung81 commented:
@fpaesani thanks for this contribution. I could build LAMMPS with the package and run the example input scripts successfully on my Linux box with GCC and MPICH. The output from the example input are consistent with the provided reference log files. I also tried running them with 4 MPI procs and these runs also give consistent results with 1 MPI proc. valgrind does not show any issues directly from the source files in the src/MBX or the MBX library.

I have a couple of questions and comments:

1) It seems that the MBX library depends on FFTW3 (per https://github.com/paesanilab/MBX). Users need to have FFTW3 installed in their environment and FFTW_HOME set to build LAMMPS with `-DPKG_MBX=on`. Is it correct? If so, could you please update the doc page for this dependency?

2) There are 2 reference log files under `mof_ff_lammps+1_h2o_mbx`, with numerical differences in `edisp` at several time steps and the [MBX] logging at the end of the run. Is there a reason to keep both?

3) Please remove commented-out codes (e.g. for debugging or unused) in `pair_mbx.cpp` and `fix_mbx.cpp`, or put them between `#ifdef MBX_DEBUG .. #endif`

Miniland1333 committed:
Delete log.03Sep25.mof.g++.1

Miniland1333 committed:
Removed unused and/or debugging code

ndtrung81 commented:
Thanks for addressing my Q2 and Q3 promptly.

Regarding Q1, is it possible to download and install FFTW3 if `DOWNLOAD_MBX` is ON and specify the installation location FFTW_HOME needed by `CONFIGURE_COMMAND`? MBX's configure command needs FFTW_HOME to point to a valid FFTW3 installation.

akohlmey commented:
> Regarding Q1, is it possible to download and install FFTW3 if `DOWNLOAD_MBX` is ON and specify the installation location FFTW_HOME needed by `CONFIGURE_COMMAND`? MBX's configure command needs FFTW_HOME to point to a valid FFTW3 installation.

That would be a bad idea leading to potential linkage conflicts when LAMMPS uses FFTW as well and uses its CMake based detection. We cannot do anything about an externally compiled library, but if the library is compiled from within the LAMMPS configuration, the CMake script should check with something like the following (off the top of my head and thus untested):
```cmake
if (FFT_SINGLE)
  message(FATAL_ERROR, "MBX package requires using double precision FFT"
endif()
find_package(FFTW3 REQUIRED)
if(FFT STREQUAL "FFTW3")
   # compile MBX using the CMake settings for FFTW
else()
  message(FATAL_ERROR, "Must compile LAMMPS with FFTW3 with MBX package")
endif()
```

LAMMPS bundles its own `cmake/Modules/FindFFTW3.cmake` file (since there is no official CMake provided one) and that sets the variables:
```cmake
#  FFTW3_INCLUDE_DIRS  - where to find fftw3.h, etc.
#  FFTW3_LIBRARIES     - List of libraries when using fftw3.
#  FFTW3_OMP_LIBRARIES - List of libraries when using fftw3.
#  FFTW3_FOUND         - True if fftw3 found.
```
Those can be used with the configure command for the downloaded source.
And for linking you can adjust the INTERFACE_LINK_LIBRARIES of the imported target accordingly.

Miniland1333 commented:
Thanks @ndtrung81 and @akohlmey for the FFTW suggestions! I'll try that out and then update the documentation accordingly.

Miniland1333 committed:
Merge remote-tracking branch 'upstream/develop' into MBX-plugin

Miniland1333 committed:
Added FFTW3_INCLUDE_DIRS to MBX plugin

Allows for MBX to use the FFTW3 found by LAMMPS and CMAKE.

Miniland1333 committed:
Update documentation1

akohlmey commented on cmake/Modules/Packages/MBX.cmake:
```diff
@@ -0,0 +1,100 @@
+if(FFTW3_FOUND)
+  list(APPEND MBX_LINK_LIBS FFTW3::FFTW3)
+else()
+  message(FATAL_ERROR, "Must compile LAMMPS with FFTW3 with MBX package")
```
@Miniland1333 This error message is not correct, but also it will never be triggered. You can just delete the "else" branch. In fact, even the "if" should not be required, but lets just keep that in case you decide to make FFTW3 optional and remove the "REQUIRED".

With `find_package(FFTW3 REQUIRED)`, CMake will stop with an error if it cannot find the development files (headers and library files) required to compile with FFTW3. This is independent from the FFT setting of LAMMPS. And you are checking this not for LAMMPS but for compiling the MBX library.

I've discussed this FFTW issue today with @sjplimp and we concluded that the check for the development files is sufficient. The potential conflicts that I am concerned about are very unlikely to happen because it is unlikely that people will use your package at the same time as PPPM.

akohlmey replied:
@ndtrung81 If you are happy with the changes, we can merge this.

Miniland1333 committed:
Removed trailing whitespace

Miniland1333 committed:
Remove redundant FFTW3 error message

This error message is not required since it would already fail at the `find_package(FFTW3 REQUIRED)` step.

Miniland1333 commented:
Okay I have removed the redundant "else" statement and removed a trailing whitespace. I see that all checks have now passed and we should be ready to merge!

I was able to compile and then run the examples as normal.

akohlmey committed:
whether LAMMPS uses single precision FFTs is irrelevant for compiling MBX

akohlmey committed:
we must use the MPI compiler wrappers or ./configure can fail

akohlmey committed:
also follow LAMMPS' settings for including OpenMP and -fPIC support (for linking a LAMMPS shared lib)

akohlmey commented:
> I see that all checks have now passed and we should be ready to merge!

Unfortunately, not yet. There are some issues that still need to be resolved to make everything work smoothly. I am working on this right now.

Please note that you are using an outdated version of the JSON library (3.7.0, current is 3.12.0 which is used internally in LAMMPS with an adjusted namespace) that will break when enabling C++20 or later. LAMMPS currently requires C++17 (and will stay at that minimum requirement for some time since we just changed it from C++11 *and* skipped over C++14), but packages may require later C++ standards. Specifically, KOKKOS has been often the package requiring a later C++ standard.

akohlmey committed:
use MD5 checksum like the rest of LAMMPS, use MBXLIB prefix instead of MBX for URL and checksum

akohlmey committed:
add missing section for PKG_MBX in doc/src/Build_extras.rst

akohlmey committed:
support building MBX as a plugin

akohlmey commented:
@Miniland1333 @cjknight @fpaesani

> Thank you for following up. If we could get the MBX integration wrap up by Jan 31, that would be awesome!
> 
> Thank you so much for your help and please let us know if there is anything we can do to facilitate this. 

I am done with some last minute changes to increase compatibility with different machines and enable building the package as a plugin (since its license conflicts with including it into distributed LAMMPS binary packages directly).

Please have a look and let me know if there is anything else that you want to change.
I am planning to merge this later today.

For some more tweaks, we can always prepare and merge a follow-up pull request. I would like to revisit the portability to Windows. At least with the Linux2Windows cross-compiler it should work. For native compilation with MSVC++ there is next to no chance because you are using autoconf. Cygwin or MinGW might work, but we don't test it anymore.

I had planned to also have a LAMMPS feature release before the end of the month, but we are currently too far behind on processing pull requests to make it. The current target data is two weeks from today on Feb 11th.

akohlmey committed:
document availability of plugin

jtclemm commented on src/MBX/README:
```diff
@@ -0,0 +1,22 @@
+This package implements the "pair_style mbx" command
+for LAMMPS, which are used to perform molecular dynamics simulations
```
"... which is used..."

Miniland1333 replied:
Fixed in 9c9da4451b4057b1a795ec62fa4b8e44e1b8f93e

jtclemm commented on doc/src/pair_mbx.rst:
```diff
@@ -0,0 +1,214 @@
+
+If you have questions not answered by this documentation, please
+reference the MBX website
+`mbxsimulations.com <https://mbxsimulations.com>`_ or reach out to us at
```
Might be more specific and clarify that us = the MBX team.

Miniland1333 replied:
Fixed in 9c9da4451b4057b1a795ec62fa4b8e44e1b8f93e

jtclemm commented on src/MBX/pair_mbx.cpp:
```diff
@@ -0,0 +1,528 @@
+
+  // energy terms available to pair compute
+
+  nextra = 22;
```
Should this be 10? I don't see anything past `pvector[9]`.

Miniland1333 replied:
Fixed in 9c9da4451b4057b1a795ec62fa4b8e44e1b8f93e. We used to have additional terms printed out when debugging.

jtclemm commented on src/MBX/fix_mbx.cpp:
```diff
@@ -0,0 +1,2126 @@
+
+    double mbx_cut = mbx_impl->ptr_mbx->GetRealspaceCutoff();
+    double diff_sq = (mbx_cut - pair_mbx->cut_global) * (mbx_cut - pair_mbx->cut_global);
+    if (diff_sq > 1e-9) error->one(FLERR, "[MBX] cutoff not consistent with LAMMPS");
```
1e-9 would be better defined as a static constexpr at the top of the file

jtclemm commented on src/MBX/fix_mbx.cpp:
```diff
@@ -0,0 +1,2126 @@
+
+void FixMBX::init()
+{
+  if (!atom->q_flag) error->all(FLERR, "[MBX] Fix mbx requires atom attribute q");
```
Should check `atom->molecule_flag` flag too.

jtclemm commented on src/MBX/fix_mbx.cpp:
```diff
@@ -0,0 +1,2126 @@
+          double dy = x[i][1] - ximage[1];
+          double dz = x[i][2] - ximage[2];
+          double rsq = dx * dx + dy * dy + dz * dz;
+          if (rsq < 0.00001) mol_local[j] = 0;
```
0.00001 would also be clearer as a static constexpr

jtclemm replied:
Would the two co-situated ghosts also not just have the same tag? Might be safer to check if their tag is the same and, if not, throw an error b/c particles otherwise should not be on top of each other.

jtclemm reviewed (approved):
Reading through the code, I added some comments. But I don't think any of them are critical per-se.

akohlmey committed:
add files for MBX plugin compilation

Miniland1333 committed:
Fixed some small changes from jtclemm

Miniland1333 commented:
Okay I have implemented the easiest of @jtclemm suggestions. I think we can merge what we have currently.


 For the remainder of @jtclemm suggestions, I'll have those tweaks in a follow-up PR. I will also update the tarball anyway in that PR once I update the JSON library in MBX to 3.12.0.

akohlmey commented:
> I'll have those tweaks in a follow-up PR.

... and I can have a second look at compiling everything for Windows with the Linux2Windows cross-compiler.

Miniland1333 committed:
Merge branch 'develop' into MBX-plugin

cjknight commented:
Great news! Thanks @akohlmey for all the great feedback and working with @Miniland1333 to improve code quality!! Thanks @ndtrung81 and @jtclemm as well for your time and contributions! We really appreciate it!

fpaesani commented:
Fantastic to see MBX now integrated into LAMMPS! Huge thanks to @akohlmey for guiding @Miniland1333 through the process and helping ensure the merge met LAMMPS standards. Thanks as well to @ndtrung81 and @jtclemm for the thoughtful reviews and feedback along the way. Much appreciated!
